cocos/2d/CCNode.h

```diff
@@ -1881,9 +1895,17 @@ class CC_DLL Node : public Ref
     mutable bool _additionalTransformDirty; ///< transform dirty ?
     bool _transformUpdated;         ///< Whether or not the Transform object was updated since the last frame
 
-    int _localZOrder;               ///< Local order (relative to its siblings) used to sort the node
+    union {
+        struct {
+            std::int32_t z; // The original localZOrder
+            std::uint32_t a; // Order Of Arrival, for avoid sort problem with unstable_sort algorithm.
+        } detail;
+        std::int64_t value; // The value to be used in sort
+    } _localZOrder;               ///< Local order (relative to its siblings) used to sort the node
```

Comment:
The codes have problem. For example, let's assume there are two nodes:
```
node1: z = 0, a = 1, then value is 0x0000000100000000
node2: z = 1, a = 0, then value is 0x0000000000000001
```
`node2.vlaue < node1.value`, then the render sequence is `node2, node1`, it is wrong. The correct sequence should be `node1, node2` because the `node2.z` is bigger. You can test `cpp-emptey-test` on Mac or iOS, you can see now the `close item` is hidden by background.

If we change it like this,
```c++
union {
        struct {
            std::int32_t z; // The original localZOrder
            std::uint32_t a; // Order Of Arrival, for avoid sort problem with unstable_sort algorithm.
        } detail;
        std::int64_t value; // The value to be used in sort
    } _localZOrder;  
```

 then it may have issue of `z` is negative, for example
```
node1: z = -1, a = 1, then value is -(2^32 + 1)
node2: z = -1, a = 2, then value is -(2^32 + 2)
```
node2.value < node1.value, so the render sequence is `node2, node1`. It is wrong, because `node2.z == node1.z` but `node2.a > node2.a`.

The above analyzation only takes effect with little endian architecture(x86, arm). If the architecture is big endian, the result is different.

So we can not just use union here. If the union is the only effect for optimization, then we should revert the PR.